authorize/evaluator/headers_evaluator.go

```diff
@@ -54,6 +57,48 @@ type HeadersResponse struct {
 	Headers http.Header
 }
 
+var variableSubstitutionFunctionRegoOption = rego.Function2(&rego.Function{
+	Name: "pomerium.variable_substitution",
+	Decl: types.NewFunction(
+		types.Args(
+			types.Named("input_string", types.S),
+			types.Named("replacements",
+				types.NewObject(nil, types.NewDynamicProperty(types.S, types.S))),
+		),
+		types.Named("output", types.S),
+	),
+}, func(bctx rego.BuiltinContext, op1 *ast.Term, op2 *ast.Term) (*ast.Term, error) {
+	inputString, ok := op1.Value.(ast.String)
+	if !ok {
+		return nil, fmt.Errorf("invalid input_string type: %T", op1.Value)
+	}
+
+	replacements, ok := op2.Value.(ast.Object)
+	if !ok {
+		return nil, fmt.Errorf("invalid replacements type: %T", op2.Value)
+	}
+
+	var err error
+	output := os.Expand(string(inputString), func(key string) string {
+		if key == "$" {
+			return "$" // allow a dollar sign to be escaped using $$
+		}
```

Comment:
I realized that someone might conceivably want to include the $ character in a header value (directly followed by some text), so I added a special case for this. This allows setting the literal value "$foo" using the syntax "$$foo" (just like how dollar signs can be escaped in Makefiles).